File: apps/ios/TaakHelden/Core/Sync/PhotoBonusService.swift
import Foundation

@Observable
final class PhotoBonusService {
    private let apiClient: TaakHeldenAPIClient

    init(apiClient: TaakHeldenAPIClient) {
        self.apiClient = apiClient
    }

    /// Uploads JPEG bytes (~2 MP target) via intent → PUT → confirm → attach.
    func uploadTaskPhoto(instanceID: String, jpegData: Data) async throws {
        let intent = try await apiClient.createUploadIntent(
            instanceID: instanceID,
            contentType: "image/jpeg",
            bytes: jpegData.count
        )

        try await apiClient.uploadPhoto(to: intent.uploadUrl, data: jpegData, contentType: "image/jpeg")
        _ = try await apiClient.confirmPhoto(photoID: intent.photoId)

        var attempts = 0
        while attempts < 10 {
            let status = try await apiClient.fetchPhotoStatus(photoID: intent.photoId)
            if status.status == "ready" {
                _ = try await apiClient.attachPhoto(instanceID: instanceID, photoID: intent.photoId)
                return
            }
            if status.status == "failed" {
                throw PhotoBonusError.processingFailed
            }
            attempts += 1
            try await Task.sleep(nanoseconds: 500_000_000)
        }

        throw PhotoBonusError.processingTimeout
    }
}

enum PhotoBonusError: LocalizedError {
    case processingFailed
    case processingTimeout

    var errorDescription: String? {
        switch self {
        case .processingFailed:
            return "Foto lukte niet — je mag het nog een keer proberen."
        case .processingTimeout:
            return "Foto wordt nagekeken… kom zo nog even terug."
        }
    }
}

File: apps/android/core/src/main/kotlin/nl/taakhelden/core/sync/PhotoBonusService.kt
package nl.taakhelden.core.sync

import kotlinx.coroutines.delay
import nl.taakhelden.core.api.TaakHeldenApiClient
import nl.taakhelden.core.i18n.UserMessage

public class PhotoBonusException(public val userMessage: UserMessage) :
    Exception(userMessage.name)

/**
 * Uploads a task photo for the photo bonus.
 *
 * The four-step dance (intent → PUT → confirm → attach) exists because the photo is only
 * attachable once the Worker has stripped EXIF and marked it `ready`; attaching earlier
 * would surface an unprocessed image to a parent.
 */
public class PhotoBonusService(
    private val apiClient: TaakHeldenApiClient,
) {
    public suspend fun uploadTaskPhoto(instanceId: String, jpegData: ByteArray) {
        val intent = apiClient.createUploadIntent(
            instanceId = instanceId,
            contentType = CONTENT_TYPE,
            bytes = jpegData.size,
        )

        apiClient.uploadPhoto(intent.uploadUrl, jpegData, CONTENT_TYPE)
        apiClient.confirmPhoto(intent.photoId)

        repeat(MAX_POLL_ATTEMPTS) {
            val status = apiClient.fetchPhotoStatus(intent.photoId)
            when (status.status) {
                "ready" -> {
                    apiClient.attachPhoto(instanceId, intent.photoId)
                    return
                }

                "failed" -> throw PhotoBonusException(UserMessage.PHOTO_PROCESSING_FAILED)
            }
            delay(POLL_INTERVAL_MS)
        }

        throw PhotoBonusException(UserMessage.PHOTO_PROCESSING_TIMEOUT)
    }

    private companion object {
        const val CONTENT_TYPE = "image/jpeg"
        const val MAX_POLL_ATTEMPTS = 10
        const val POLL_INTERVAL_MS = 500L
    }
}
